src/main/scala/rocketchip/Passes.scala

```diff
@@ -0,0 +1,78 @@
+// See LICENSE for license details.
+
+package rocketchip
+
+import firrtl._
+import firrtl.ir._
+import firrtl.passes.Pass
+
+// Converts some modules to external modules, based on a given function.  If
+// that function returns "true" then the module is converted into an ExtModule,
+// otherwise it's left alone.
+class ConvertToExtMod(classify: (Module) => Boolean) extends Pass {
+  def name = "Convert to External Modules"
+
+  def run(c: Circuit): Circuit = {
+    val modulesx = c.modules.map {
+      case m: ExtModule => m
+      case m: Module =>
+        if (classify(m)) {
+          new ExtModule(m.info, m.name, m.ports)
+        } else {
+          m
+        }
+    }
+    Circuit(c.info, modulesx, c.main)
+  }
+}
+
+// "Re-Parents" a circuit, which changes the top module to something else.
+class ReParentCircuit(newTopName: String) extends Pass {
+  def name = "Re-Parent Circuit"
+
+  def run(c: Circuit): Circuit = {
+    Circuit(c.info, c.modules, newTopName)
+  }
+}
+
+// Removes all the unused modules in a circuit by recursing through every
+// instance (starting at the main module)
+class RemoveUnusedModules extends Pass {
```

Comment:
I'm not familiar enough with firrtl passes to review this, @azidar @jackkoenig can you check this out?